subsys/bluetooth/host/att.c

```diff
@@ -198,10 +198,23 @@ static int chan_send(struct bt_att_chan *chan, struct net_buf *buf,
 		}
 	}
 
+	net_buf_simple_save(&buf->b, &state);
+
 	chan->sent = cb ? cb : chan_cb(buf);
 
-	return bt_l2cap_send_cb(chan->att->conn, BT_L2CAP_CID_ATT, buf,
-				att_cb(chan->sent), &chan->chan.chan);
+	/* Take a ref since bt_l2cap_send_cb takes ownership of the buffer */
+	err = bt_l2cap_send_cb(chan->att->conn, BT_L2CAP_CID_ATT,
+				net_buf_ref(buf), att_cb(chan->sent),
+				&chan->chan.chan);
+	if (!err) {
+		net_buf_unref(buf);
+		return 0;
+	}
+
+	net_buf_simple_restore(&buf->b, &state);
```

Comment:
Seems fine now, with one exception. I am looking at `chan_req_send()` and this fragment:
```
	/* Keep a reference for resending in case of an error */
	err = chan_send(chan, net_buf_ref(req->buf), NULL);
	if (err) {
		net_buf_unref(req->buf);
		chan->req = NULL;
	}
```
seems suspicious to me. First of all "keep reference for resending in case of error" does not match what the code does: buffer will be referenced first, then unreferenced when error occurs - as a result buffer will be freed. Additionally properly sent buffer will be leaked, as all `chan_req_send` callers expect buffer to be freed/consumed.

Reply:
The resending logic it mentions is for the request itself, in case it is not for chan_send, the buffer is release at att_handle_rsp, we wouldn't keep a reference in bt_att_req if we expected it to be consumed. That said the error path seems wrong now, it would not free the buffer just the extra reference, what we should do is only take a reference if chan_send succeeds.

Reply:
Actually we always do prepend it back to att->reqs in case of an error, so it looks like it was broken already as release the reference and reset the req->buf to NULL, weird that we didn't hit this before though.

Reply:
Ive end up fixing the behavior of chan_req_send so it doesn't reset the req->buf to NULL.